test/cpp/end2end/connection_delay_injector.cc

```diff
@@ -0,0 +1,116 @@
+// Copyright 2016 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "test/cpp/end2end/connection_delay_injector.h"
+
+#include <atomic>
+#include <memory>
+
+#include "absl/memory/memory.h"
+#include "absl/utility/utility.h"
+
+#include "src/core/lib/channel/channel_args.h"
+#include "src/core/lib/gprpp/time.h"
+#include "src/core/lib/iomgr/tcp_client.h"
+#include "src/core/lib/iomgr/timer.h"
+
+// defined in tcp_client.cc
+extern grpc_tcp_client_vtable* grpc_tcp_client_impl;
+
+namespace grpc {
+namespace testing {
+
+namespace {
+
+grpc_tcp_client_vtable* g_original_vtable = nullptr;
+
+std::atomic<grpc_core::Duration> g_delay;
+
+class InjectedDelay {
+ public:
+  InjectedDelay(grpc_closure* closure, grpc_endpoint** ep,
+                grpc_pollset_set* interested_parties,
+                const grpc_channel_args* channel_args,
+                const grpc_resolved_address* addr,
+                grpc_core::Timestamp deadline)
+      : closure_(closure),
+        endpoint_(ep),
+        interested_parties_(interested_parties),
+        channel_args_(grpc_channel_args_copy(channel_args)),
+        deadline_(deadline) {
+    memcpy(&address_, addr, sizeof(grpc_resolved_address));
+    GRPC_CLOSURE_INIT(&timer_callback_, TimerCallback, this, nullptr);
+    auto duration = g_delay.load();
+    deadline_ += duration;
+    grpc_timer_init(&timer_, grpc_core::ExecCtx::Get()->Now() + duration,
+                    &timer_callback_);
+  }
+
+  ~InjectedDelay() { grpc_channel_args_destroy(channel_args_); }
+
+ private:
+  static void TimerCallback(void* arg, grpc_error_handle /*error*/) {
+    auto* self = static_cast<InjectedDelay*>(arg);
+    g_original_vtable->connect(self->closure_, self->endpoint_,
+                               self->interested_parties_, self->channel_args_,
+                               &self->address_, self->deadline_);
```

Comment:
Can we subtract the initial delay from the delay used here? Thinking without that, this may cause surprising behavior

Reply:
We account for that on line 55 above.  This is basically what the original code in client_lb_end2end_test does.

Reply:
For least surprising behavior, though, shouldn't line 55 really be removed?

Also shouldn't `duration` be set to something like `max(duration, deadline - Now())` ?

Otherwise, if subchannel connection timeout is e.g. 20 seconds, and we inject a 30 second delay, the subchannel won't actually receive it's timeout until 30 seconds later?

Reply:
Good point.  I had thought there must have been some reason the existing code worked this way, but I traced it back to #13494, and I don't see anything in that (long) review that explains why it's done this way.  And the tests do seem to pass without line 55.  I'll remove that line in a subsequent PR.

I don't think we need to subtract from deadline, because the deadline is a timestamp, not a duration.  If the delay is longer than `deadline - Now()`, then deadline will be in the past when the connection attempt starts, in which case it will fail immediately.

Reply:
Oh wait, sorry, you were saying the opposite: you were saying we should cap the delay by the deadline.  Yeah, that's probably a good idea.  I'll make that change as well.

Reply:
Sent you #29349 with this change.